Add an Automattic abstract address autocomplete provider

senadir opened the pull request:
### Changes proposed in this Pull Request:

This PR introduces an abstract address provider class and JS script, the goal of this class is that it's consumed directly by other services, ensuring they do the minimal amount of work needed, which is:
- Provide a name and ID.
- Provide a callback function that does the JWT token fetching, depending on that plugin system, secrets, routes, and whatever.
The rest is fully handled by the abstract provider and by the frontend script, including error handling, debouncing, caching, tracking, and auto disabling.

Once an address selection is done, the duration of all requests is collected and submitted to statsd to keep count of it and optimize server setup over time.

An Automattic service would need the following code:

1. A class that extends `AbstractAutomatticAddressProvider`
```php
<?php

declare( strict_types=1 );

use Automattic\WooCommerce\Internal\AddressProvider\AbstractAutomatticAddressProvider;

/**
 * Mock Address Provider
 */
class AutomatticAddressProvider extends AbstractAutomatticAddressProvider {

	/**
	 * Constructor.
	 */
	public function __construct() {
		$this->id   = 'automattic-address-provider';
		$this->name = 'Automattic Address Provider';
		parent::__construct();
	}

	public function get_address_service_jwt() {
		// Fetch token here, cache and retries are handled in `AbstractAutomatticAddressProvider`.
		return "TOKEN";
	}

	public function can_telemetry() {
		// Check if this plugin is allowed to send telemetry data to Automattic.
		return true;
	}
}
```

And need to register said class:

```php
add_action( 'woocommerce_loaded', 'automattic_address_provider_init' );

/**
 * Initialize the mock address provider.
 *
 * @return void
 */
function automattic_address_provider_init() {
	include_once __DIR__ . '/AutomatticAddressProvider.php';
	add_filter(
		'woocommerce_address_providers',
		function ( $providers ) {
			$providers[] = 'AutomatticAddressProvider';
			return $providers;
		}
	);
}
```

### Screenshots or screen recordings:

https://github.com/user-attachments/assets/b6ded7ff-ce2f-445b-8226-55b69f48638d

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:
0. This feature is still behind experimental flag, so you need to be watching, not just building.
1. Install the following 2 plugins and enable them, [automattic-address-provider-1.zip](https://github.com/user-attachments/files/21165290/automattic-address-provider-1.zip) and [automattic-address-provider-2.zip](https://github.com/user-attachments/files/21165291/automattic-address-provider-2.zip). For all purposes, they're similar, just with different IDs, and are meant to simulate having 2 plugins enabled at the same time. The tokens in those plugins are modified to last 30 days, if you need a new token, reach out to @opr or @senadir, or follow steps from 1 to 11 [in this PR.](https://github.a8c.com/Automattic/wpcom/pull/185205)
2. For all purposes, this PR focuses on the provider aspect and getting back results, so validate you're getting correct results, and that they match what you're searching for and that UX is good, things like the accessibility and interactions with the actual field will be part of a separate PR.
3. On Checkout, select a country and start typing your address.
4. Ensure that addresses start showing up after the third letter, with a little debounce.
5. Select an address, ensure values are filled.
6. Do the same with the shipping form, but for a different country.
7. Open console, ensure there are no errors coming from the service.
8. While opening console, select an address, make sure there's a request to `boom.gif`, this might be blocked by tracking blockers.
9. Visit https://mc.a8c.com/grafana/d/ceqwhp6j8twxsf/address-autocomplete-service?orgId=1&from=now-30m&to=now&timezone=utc and see you requests there, including how many requests you made and their average time.
<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Add implematnion
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

opr committed:
Create address autocomplete registration JS function

opr committed:
Register and enqueue address autocomplete script, add provider data too

opr committed:
Add unit tests to cover legacy js functionality

opr committed:
Add better checks for enqueuing autocomplete script

opr committed:
Add styling for suggestions list

opr committed:
Determine browser when loading the page, used later for autofill

senadir committed:
rebase

senadir committed:
commit pending work

senadir committed:
fix the test

senadir committed:
Fix debounce, remove signal, add statd telementry

senadir commented on plugins/woocommerce/client/legacy/css/address-autocomplete.scss:
Just prettier changes

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+( function () {
+	const permanentlyDisabledServices = [];
```
I keep count of services disabled here, just in case we have a broken thing, it just auto skip itself in canSearch

senadir replied:
Disabling is per client session, if you refresh the page, the limits are gone.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+	const searchUrl = `${ baseUrl }/search`;
+	const selectUrl = `${ baseUrl }/select`;
+
+	const MAX_SERVICE_ERROR_RETRIES = 3;
```
After getting 3 5xx errors, we just disable the provider.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+	const MAX_SERVICE_ERROR_RETRIES = 3;
+
+	/**
+	 * Debounce function from lodash, modified to return a promise.
```
I needed a good debounce function, I'm not aware of a way to include npm modules in this scripting world, so I'm just copy pasting the whole function, with some modification, you can see a diff here https://www.diffchecker.com/TINeimGv/

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+		return debounced;
+	}
+
+	Object.entries( a8cAddressAutocompleteServiceKeys ).forEach(
```
This script will be enqueued once for all providers, they all just into this global var and add their tokens.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+
+	Object.entries( a8cAddressAutocompleteServiceKeys ).forEach(
+		( [ key, value ] ) => {
+			let sessionId =
```
Each session need a session ID that we generate here via the browser.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+					} );
+
+					try {
+						const startTime = Date.now();
+						const response = await fetch(
+							`${ searchUrl }?${ params.toString() }`
+						);
+						const endTime = Date.now();
+						requestDurations.push( endTime - startTime );
+						let data = await response.json();
```
Keeping count of timing to track it later.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+						requestDurations.push( endTime - startTime );
+						let data = await response.json();
+
+						if ( data.code ) {
+							switch ( data.code ) {
+								case 'expired_jwt_token':
+								case 'malformed_jwt_token':
+								case 'invalid_jwt_token':
+								case 'invalid_issuer':
+								case 'invalid_service':
+								case 'missing_jwt_token':
+									permanentlyDisabledServices.push( key );
+									console.error(
+										'Automattic Address Suggestion has been disabled due to invalid JWT token'
+									);
+									return [];
+								case 'rate_limit_exceeded':
+									permanentlyDisabledServices.push( key );
+									setTimeout( () => {
+										permanentlyDisabledServices.splice(
+											permanentlyDisabledServices.indexOf(
+												key
+											),
+											1
+										);
+									}, response.headers.get( 'RateLimit-Retry-After' ) * 1000 );
+									console.error(
+										'Automattic Address Suggestion has been disabled due to rate limit exceeded'
+									);
+									return [];
```
This error handling is more or less the same between this and select, except what to return and a couple of unique errors per endpoint, maybe there's a chance to move it later.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+											),
+											1
+										);
+									}, response.headers.get( 'RateLimit-Retry-After' ) * 1000 );
```
If you get rate limited, we wait until the rate limit is off.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+								case 'invalid_issuer':
+								case 'invalid_service':
+								case 'missing_jwt_token':
+									permanentlyDisabledServices.push( key );
```
A malformed token can't be solved with another request, so we just disable.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+			);
+			window.wc.addressAutocomplete.registerAddressAutocompleteProvider( {
+				id: key,
+				canSearch: () => {
+					try {
+						if ( permanentlyDisabledServices.includes( key ) ) {
+							return false;
+						}
+
+						// Split JWT into parts
+						const [ , payload ] = value.split( '.' );
+						if ( ! payload ) {
+							permanentlyDisabledServices.push( key );
+							return false;
+						}
+
+						// Decode payload
+						const decodedPayload = JSON.parse( atob( payload ) );
+
+						// Check expiration
+						const currentTime = Math.floor( Date.now() / 1000 );
+						if (
+							! decodedPayload.exp ||
+							decodedPayload.exp < currentTime
+						) {
+							permanentlyDisabledServices.push( key );
+							return false;
+						}
+
+						return true;
```
Can search validate if the token we have is valid, we don't check country here, maybe we should make this extendible at some point.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+
+					let data = await response.json();
+					// Reset session ID after successful select
+					sessionId = crypto.randomUUID();
+					try {
+						dispatchEvent(
+							new CustomEvent(
+								'wc-address-autocomplete-service-request-durations',
+								{
+									detail: {
+										requestDurations,
+									},
+								}
+							)
+						);
```
Upon selection we do trigger this event to be consumed by our perf tracking logic.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,485 @@
+	);
+} )();
+
+function createBeacon( section, { name, value, type } ) {
+	const event = name.replace( '-', '_' );
+
+	// A counting event defaults to incrementing by one.
+	if ( type === 'counting' ) {
+		value = value === undefined ? 1 : value;
+	}
+	value = Array.isArray( value ) ? value : [ value ];
+	return value.map(
+		( v ) =>
+			`a8c.${ section }.${ event }:${ v }|${
+				type === 'timing' ? 'ms' : 'c'
+			}`
+	);
+}
+
+function createStatsdURL( sectionName, events ) {
+	if ( ! Array.isArray( events ) ) {
+		events = [ events ]; // Only a single event was passed to process.
+	}
+
+	const sanitizedSection = sectionName.replace( /[.:-]/g, '_' );
+	const json = JSON.stringify( {
+		beacons: events
+			.map( ( event ) => createBeacon( sanitizedSection, event ) )
+			.flat(),
+	} );
+
+	const encodedJson = encodeURIComponent( json );
+
+	return `https://pixel.wp.com/boom.gif?json=${ encodedJson }`;
+}
```
This code is copied and modified from a calypso code, here we use Automattic's statsd library and we send several beacons, each with the time we had.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/address-autocomplete.js:
```diff
@@ -663,7 +662,7 @@ window.wc.addressAutocomplete.registerAddressAutocompleteProvider =
 					);
 				}
-			}, 200 );
+			}, 50 );
```
Reduced the timeout needed to fill values from select, honestly even 50 is too much, we just need to wait for states to load in this.

senadir commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,165 @@
+	 * Loads up the JWT for the address service and saves it to transient.
+	 */
+	public function __construct() {
+		add_filter( 'pre_update_option_woocommerce_address_autocomplete_enabled', array( $this, 'refresh_cache' ) );
```
Every time the address autocomplete feature get enabled or disabled, we refresh the cache, just in case we have malformed cache.

senadir commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,165 @@
+
+		$transient_key = $this->id . 'address_autocomplete_jwt';
+		$cached_jwt    = get_transient( $transient_key );
+		// If we have a cached, valid token, we load it to class and return early.
+		if ( $cached_jwt && JsonWebToken::shallow_validate( $cached_jwt ) && 'local' !== wp_get_environment_type() ) {
+			$this->jwt = $cached_jwt;
+			return $this->jwt;
+		}
```
Skip cache during dev to make switching tokens easier.

senadir commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,165 @@
+			$fresh_jwt = $this->get_address_service_jwt();
+			if ( $fresh_jwt && JsonWebToken::shallow_validate( $fresh_jwt ) ) {
+				$this->jwt = $fresh_jwt;
+				set_transient( $transient_key, $fresh_jwt, DAY_IN_SECONDS );
```
A token is valid for one day, I think we can maybe change this so the token is value for its expiry date.

senadir commented on plugins/woocommerce/src/StoreApi/Utilities/JsonWebToken.php:
```diff
@@ -50,6 +50,30 @@ public static function create( array $payload, string $secret ) {
+	 *
+	 * @return bool
+	 */
+	public static function shallow_validate( string $token ) {
```
I needed a way to validate that a token is structurally valid without checking the secret part, which is fine.

senadir committed:
Update plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

senadir committed:
fix typos

senadir committed:
add PR self feedback

senadir committed:
only send events if the provider can actually send events

senadir committed:
fix typo

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+						return [];
+					}
+
+					inputValue = inputValue.trim();
```
The input is already sanitized in the address-autcomplete.js place before being passed to us, trimming here is done to skip requests where only a space was added.

senadir commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+							return false;
+						}
+
+						// Split JWT into parts
+						const [ , payload ] = value.key.split( '.' );
+						if ( ! payload ) {
+							permanentlyDisabledServices.push( key );
+							return false;
+						}
+
+						// Decode payload
+						const decodedPayload = JSON.parse( atob( payload ) );
+
+						// Check expiration
+						const currentTime = Math.floor( Date.now() / 1000 );
+						if (
+							! decodedPayload.exp ||
+							decodedPayload.exp < currentTime
+						) {
+							permanentlyDisabledServices.push( key );
+							return false;
+						}
+
+						return true;
+					} catch ( e ) {
+						permanentlyDisabledServices.push( key );
+						return false;
+					}
```
Proper JWT validation is done at the API level, here we only check if it's expired or not.

straku commented on plugins/woocommerce/changelog/59426-add-abstract-automattic-provider:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: add
+
+Add implematnion
```
```suggestion
Add implementation
```
There's a typo. But I would also suggest making it a bit less vague.

senadir replied:
```suggestion
Add Automattic abstract address provider
```

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+					const entries = Array.from( searchCache.entries() );
+					entries.slice( 0, 20 ).forEach( ( [ key ] ) => {
+						searchCache.delete( key );
+					} );
```
```suggestion
					const keys = Array.from( searchCache.keys() );
					keys.slice( 0, 20 ).forEach( ( [ key ] ) => {
						searchCache.delete( key );
					} );
```
Looks like you only use keys here.

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+
+				const errorCode = data.code || data.error;
+
+				switch ( errorCode ) {
```
Leaving one comment, but it applies to several `console.error` logs here. Should we also log the key of the service alongside the message?

senadir replied:
Sure thing!

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+						permanentlyDisabledServices.push( key );
+						setTimeout( () => {
+							permanentlyDisabledServices.splice(
+								permanentlyDisabledServices.indexOf( key ),
```
Low chance, but this can break if `indexOf` returns -1. Splice with -1 will remove the last element of the array.

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+								permanentlyDisabledServices.indexOf( key ),
+								1
+							);
+						}, response.headers.get( 'RateLimit-Retry-After' ) * 1000 );
```
The standard `Retry-After` header can provide a date to retry after too, not only seconds. Can you tell me a bit more about `RateLimit-Retry-After`? Can we trust for it to always return seconds?

senadir replied:
Yeah we can, I implemented it. https://github.a8c.com/Automattic/wpcom/blob/trunk/wp-content/rest-api-plugins/endpoints/woo-address-autocomplete-endpoint.php#L504

straku replied:
😎

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+						sessionId =
+							crypto && crypto.randomUUID
+								? crypto.randomUUID()
+								: Math.random().toString( 36 ).substring( 2 );
```
It looks like we are creating `sessionId` synchronously at the beginning of processing, so I don't see an option for it to be useful here 🤔 If it is, I would love to learn how and suggest extracting it to a function, as we repeat it in two places and the implementation is pretty specific.

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+					const params = new URLSearchParams( {
+						query: inputValue,
+						country,
+						lang: document.documentElement.lang,
```
```suggestion
						lang: document.documentElement.lang || navigator.lang,
```
Maybe worth doing just in case?

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+					} );
+
+					try {
+						const startTime = Date.now();
```
We can consider tracking those with stuff available in [`Performance`](https://developer.mozilla.org/en-US/docs/Web/API/Performance) interface

senadir replied:
Sure thing

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+						const errorResult = handleApiError( data, response );
+						if ( errorResult !== null ) {
+							return errorResult;
+						}
```
Looking at `handleApiError` this doesn't seem possible? It will always return `null`. Maybe it shouldn't have a return value at all and only deal with side effects?

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,459 @@
+					let data = await response.json();
+					// Reset session ID after successful select
+					sessionId =
+						crypto && crypto.randomUUID
+							? crypto.randomUUID()
+							: Math.random().toString( 36 ).substring( 2 );
```
Let's definitely extract it to a function (3rd time the charm 😊)

straku commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,191 @@
+			}
+		}
+
+		return DAY_IN_SECONDS;
```
Is it a safe default for invalid token?

senadir replied:
You raise a good point, we should have some sort of retry mechanism.

straku commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,191 @@
+		if ( JsonWebToken::shallow_validate( $jwt ) ) {
+			$parts = JsonWebToken::get_parts( $jwt );
+			if ( property_exists( $parts->payload, 'exp' ) ) {
+				return $parts->payload->exp - time();
```
Noting that this can be negative

straku commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,191 @@
+		$suffix  = Constants::is_true( 'SCRIPT_DEBUG' ) ? '' : '.min';
+		$version = Constants::get_constant( 'WC_VERSION' );
+
+		if ( ! wp_script_is( 'a8c-address-autocomplete-service', 'registred' ) ) {
```
```suggestion
		if ( ! wp_script_is( 'a8c-address-autocomplete-service', 'registered' ) ) {
```

straku commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,191 @@
+
+
+	/**
+	 * Enqueues the checkout script, checks if it's already registered or not so we don't dubplicate, and prints out the JWT to the page to be consumed.
```
```suggestion
	 * Enqueues the checkout script, checks if it's already registered or not so we don't duplicate, and prints out the JWT to the page to be consumed.
```

straku commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,191 @@
+			'a8c-address-autocomplete-service',
+			sprintf(
+				'var a8cAddressAutocompleteServiceKeys = a8cAddressAutocompleteServiceKeys || {}; a8cAddressAutocompleteServiceKeys[ "%1$s" ] = { key: %2$s, canTelemetry: %3$s };',
+				$this->id,
```
```suggestion
				wp_json_encode( $this->id ),
```
I am no expert, got some help with Claude 😅

straku commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,191 @@
+				'var a8cAddressAutocompleteServiceKeys = a8cAddressAutocompleteServiceKeys || {}; a8cAddressAutocompleteServiceKeys[ "%1$s" ] = { key: %2$s, canTelemetry: %3$s };',
+				$this->id,
+				wp_json_encode( $this->get_jwt() ),
+				wp_json_encode( false !== $this->can_telemetry() && (bool) $this->can_telemetry() )
```
```suggestion
				wp_json_encode( $this->can_telemetry() )
```
Can we do this? If not what am I missing?

senadir replied:
We don't know what `can_telemetry` return, it can be an error or something, so I'm trying to be defensive.

straku replied:
Makes sense 👍

straku reviewed (changes requested):
Finished my initial pass over the code. Not all, but some of the comments need to be addressed. I will test it out manually tomorrow morning. Nice job 😌

senadir committed:
Apply suggestions from code review

Co-authored-by: Łukasz Strączyński <8038404+straku@users.noreply.github.com>

senadir committed:
Apply suggestions from code review

Co-authored-by: Łukasz Strączyński <8038404+straku@users.noreply.github.com>

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

senadir committed:
Address feedback

senadir committed:
add phpunit tests

senadir commented:
Thank you @straku I addressed the review and added tests and switched the cache system from transients to options, matching what WooPayments and Stripe do, because `transients` are unreliable and sometimes doens't work at all.

senadir committed:
Update plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

senadir committed:
fix typo

senadir committed:
fix bug in which enqueuening scripts would fail

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,455 @@
+			let requestDurations = [];
+			let serviceErrorRetries = 0;
+			// Cache for search results - key: `${inputValue}:${country}`, value: data
+			const searchCache = new Map();
```
Note for the future: it would be nice to use a proper LRU cache. Removing the 20 latest elements from the cache is fine to keep its size in check, but we can be removing frequently used items (not really a deal breaker since this is scoped to user session, so it might be a premature optimization).

senadir replied:
Your wish is some AI's command.

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,455 @@
+							if ( index !== -1 ) {
+								permanentlyDisabledServices.splice( index, 1 );
+							}
+						}, response.headers.get( 'RateLimit-Retry-After' ) * 1000 );
```
I still feel a bit of unease looking at this line. I think because of type mismatch, we are trusting that this string will be a number representing a second, but this is an assumption that you don't understand just by looking at this code.

Maybe worth parsing the number directly and checking for `NaN` before multiplication and usage in `setTimeout`?

senadir replied:
Sure thing yeah

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,455 @@
+				},
+			} );
+
+			window.addEventListener(
```
Should we pull that out from the loop? What's the reason for registering this event listener for each service?

senadir replied:
We don't know which service would be active at that time to send the event.

straku replied:
You could store the functions to send telemetry data in an array and then loop through them when the event happens. I think this should work? It's also not major, as I don't believe we will be dealing with many services.

senadir replied:
I think it's fine, we also need to check if this one can do telemetry. We would end up loop over them anyway; better to have a single loop

straku commented on plugins/woocommerce/client/legacy/js/frontend/a8c-address-autocomplete-service.js:
```diff
@@ -0,0 +1,455 @@
+		return debounced;
+	}
+
+	Object.entries( a8cAddressAutocompleteServiceKeys ).forEach(
```
Can we pull out util functions from the loop?

senadir replied:
Which one would that be, the ones inside the loop are dependant on the loop or would change from one to another

straku replied:
Took a deeper look and can't spot any. I initially thought that the `handleApiError`  could be extracted, but it sets missing `sessionId`. Sorry for spaming then!

senadir commented on plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php:
```diff
@@ -0,0 +1,275 @@
+	 *
+	 * @param string $jwt The JWT for the address service.
+	 */
+	public function set_jwt( $jwt ) {
+		$this->jwt = $jwt;
+		if ( null !== $jwt ) {
+			$cache_duration = $this->get_jwt_cache_duration( $jwt );
+			// If the token is expired, we don't cache it and we fetch a new one.
+			if ( 0 === $cache_duration ) {
+				$this->jwt = null;
+				$this->load_jwt();
+				return;
+			}
+			$this->update_cached_option( 'address_autocomplete_jwt', $jwt, $cache_duration );
+		} else {
+			$this->delete_cached_option( 'address_autocomplete_jwt' );
+		}
+	}
```
Shouldn't this fail at checking for last fetch attempt?

straku commented:
Notes from my testing:
- There appears to be a problem with selecting an address that contains an apartment number. When I searched for "al pokoju 4/2" (in Poland) I was able to get a reliable 500 response every time I tried to select an address.
  - Sometimes it works, but the apartment is not preserved in the selection. So searching for "Chopina 33/2" gets a suggestion called "Chopina 33/2" but when I click on it, the address is "Chopina 33".
  - I don't think it's a problem with this PR, but the API, but still letting know about it to get it tracked and fixed.
- I am not getting requests to `boom.gif` and not getting my requests into Grafana. I've tried in Vivaldi with disabled tracking and Chrome. The result is the same.
- Otherwise I tried with priority set to both providers, and I must say the flow is very smooth 👏

@senadir, before approving I think you should look into the issue with tracking (or please let me know what I am doing wrong).

straku commented on plugins/woocommerce/tests/php/src/Internal/AddressProvider/AbstractAutomatticAddressProviderTest.php:
```diff
@@ -0,0 +1,647 @@
+			->expects( $this->once() )
+			->method( 'error' )
+			->with(
+				$this->stringContains( 'Failed loding JWT for Test Provider address autocomplete service with error Test exception.' ),
```
```suggestion
				$this->stringContains( 'Failed loading JWT for Test Provider address autocomplete service with error Test exception.' ),
```

straku commented on plugins/woocommerce/tests/php/src/Internal/AddressProvider/AbstractAutomatticAddressProviderTest.php:
```diff
@@ -0,0 +1,647 @@
+		// Call load_scripts.
+		$provider->load_scripts();
+
+				// Check if inline script was added.
```
```suggestion
		// Check if inline script was added.
```

straku commented on plugins/woocommerce/tests/php/src/Internal/AddressProvider/AbstractAutomatticAddressProviderTest.php:
```diff
@@ -0,0 +1,647 @@
+		// Call load_scripts.
+		$provider->load_scripts();
+
+				// Check if inline script was added with null JWT.
```
```suggestion
		// Check if inline script was added with null JWT.
```

senadir committed:
add backoff retries to fetching the key

senadir committed:
Move to LRU cache, add backoff retries and test that

senadir commented on plugins/woocommerce/tests/php/src/Internal/AddressProvider/AbstractAutomatticAddressProviderTest.php:
```diff
@@ -0,0 +1,861 @@
+			}
+
+			// Create a new mock logger for each test case to avoid conflicts.
+			$mock_logger = $this->getMockBuilder( 'WC_Logger_Interface' )->getMock();
+			$mock_logger
+				->expects( $this->once() )
+				->method( 'error' )
+				->with(
+					$this->stringContains( sprintf( 'Failed loading JWT for Test Provider address autocomplete service (attempt %d) with error Test failure.', $test_case['expected_attempt'] ) ),
+					'address-autocomplete'
+				);
+
+			// Temporarily override the logger for this test case.
+			add_filter(
+				'woocommerce_logging_class',
+				function () use ( $mock_logger ) {
+					return $mock_logger;
+				}
+			);
```
Sadly we can't because we have several calls for mocking here so we need to redo it, we're also removing the filter.

senadir committed:
dont try to unwrap again

straku reviewed (approved):
Tracking works now for me with the second provider that you attached in the PR description. I can see the request and data in grafana.

@senadir Did you create a ticket for the 500 responses that I got when selecting an address with an apartment number? Or do you have an open PR for that? I just want to close this loop 😊

senadir commented:
Yes, here's a PR that fixes it
https://github.a8c.com/Automattic/wpcom/pull/187510
